File: LinkedLists/MyDoublyLinkedList.java
package LinkedLists;

public class MyDoublyLinkedList {
  private static class ListNodeDouble {
    int val;
    ListNodeDouble prev;
    ListNodeDouble next;

    ListNodeDouble(int val) {
      this.val = val;
      this.prev = null;
      this.next = null;
    }

    ListNodeDouble(int val, ListNodeDouble prev, ListNodeDouble next) {
      this.val = val;
      this.prev = prev;
      this.next = next;
    }
  }

  ListNodeDouble head;
  ListNodeDouble tail;
  int size;

  public MyDoublyLinkedList() {
    // Use dummy head and tail
    this.head = new ListNodeDouble(0);
    this.tail = new ListNodeDouble(0, head, null);
    this.size = 0;
    head.prev = null;
    head.next = tail;
  }

  public int get(int index) {
    if (index < 0 || index >= size) {
      return -1;
    }

    ListNodeDouble node = getNodeAtIndex(index);
    return node.val;
  }

  public void addAtHead(int val) {
    addAtIndex(0, val);
  }

  public void addAtTail(int val) {
    addAtIndex(size, val);
  }

  public void addAtIndex(int index, int val) {
    if (index < 0 || index > size) {
      return;
    }

    ListNodeDouble current = getNodeAtIndex(index);
    ListNodeDouble newNode = new ListNodeDouble(val, current.prev, current);
    current.prev.next = newNode;
    current.prev = newNode;
    size++;
  }

  public void deleteAtIndex(int index) {
    if (index < 0 || index >= size) {
      return;
    }

    ListNodeDouble nodeToDelete = getNodeAtIndex(index);
    nodeToDelete.prev.next = nodeToDelete.next;
    nodeToDelete.next.prev = nodeToDelete.prev;
    size--;
  }

  private ListNodeDouble getNodeAtIndex(int index) {
    ListNodeDouble current;
    if (index <= size / 2) {
      current = head.next;
      for (int i = 0; i < index; i++) {
        current = current.next;
      }
    } else {
      current = tail;
      for (int i = 0; i < size - index; i++) {
        current = current.prev;
      }
    }
    return current;
  }
}

// Time Complexity: O(n) for get, addAtIndex, deleteAtIndex; O(1) for addAtHead and addAtTail

File: LinkedLists/myDoublyLinkedList.js
class ListNode {
  constructor(val, prev = null, next = null) {
    this.val = val;
    this.prev = prev;
    this.next = next;
  }
}

class MyDoublyLinkedList {
  constructor() {
    // Use dummy head and tail nodes to simplify edge cases
    this.head = new ListNode(0);
    this.tail = new ListNode(0, this.head);
    this.size = 0;
    this.head.next = this.tail;
  }

  get(index) {
    if (index < 0 || index >= this.size) return -1;
    const node = this.#getNodeByIndex(index);
    return node.val;
  }

  addAtHead(val) {
    this.addAtIndex(0, val);
  }

  addAtTail(val) {
    this.addAtIndex(this.size, val);
  }

  addAtIndex(index, val) {
    if (index < 0 || index > this.size) return;
    const nodeAtIndex = this.#getNodeByIndex(index);
    const newNode = new ListNode(val, nodeAtIndex.prev, nodeAtIndex);
    newNode.prev.next = newNode;
    nodeAtIndex.prev = newNode;
    this.size++;
  }

  deleteAtIndex(index) {
    if (index < 0 || index >= this.size) return;
    const nodeToDelete = this.#getNodeByIndex(index);
    nodeToDelete.prev.next = nodeToDelete.next;
    nodeToDelete.next.prev = nodeToDelete.prev;
    this.size--;
  }

  #getNodeByIndex(index) {
    let current;
    if (index <= this.size / 2) {
      // Start at this.head.next because index=0 should return actual head
      current = this.head.next;
      for (let i = 0; i < index; i++) {
        current = current.next;
      }
    } else {
      // Start at this.tail because index=this.size should return dummy tail
      current = this.tail;
      for (let i = 0; i < this.size - index; i++) {
        current = current.prev;
      }
    }

    return current;
  }
}

// Time: O(n) for get, addAtIndex, and deleteAtIndex, O(1) for addAtHead and addAtTail
// Space: O(n) for the linked list itself
